src/impl/vamp/planning/rrt_star.hh

```diff
@@ -0,0 +1,285 @@
+#include <iostream>
+
+namespace vamp::planning
+{
+    template <typename Robot, typename RNG, std::size_t rake, std::size_t resolution>
+    struct RRT_star
+    {
+        using Configuration = typename Robot::Configuration;
+        static constexpr auto dimension = Robot::dimension;
+        static constexpr auto space_measure = Robot::space_measure;
+
+        inline static auto solve(
+            const Configuration &start,
+            const Configuration &goal,
+            const collision::Environment<FloatVector<rake>> &environment,
+            const RRT_star_settings &settings) noexcept -> PlanningResult<dimension>
+        {
+            return solve(start, std::vector<Configuration>{goal}, environment, settings);
+        }
+
+        inline static auto solve(
+            const Configuration &start,
+            const std::vector<Configuration> &goals,
+            const collision::Environment<FloatVector<rake>> &environment,
+            const RRT_star_settings &settings) -> PlanningResult<dimension>
+        {
+            // algorithm specific constants
+            static const float unit_ball_volume =
+                std::pow(std::sqrt(M_PI), dimension) / std::tgamma(dimension / 2.0 + 1.0);
+            static const float free_volume = space_measure();
+            static const float dim_recip = 1.0 / dimension;
+            static const float gamma_rrt =
+                std::pow(2 * (1.0 + 1.0 / dimension) * (free_volume / unit_ball_volume), dim_recip);
+
+            PlanningResult<dimension> result;
+            NN<dimension> tree;
+            constexpr const std::size_t start_index = 0;
+
+            auto buffer = std::unique_ptr<float, decltype(&free)>(
+                vamp::utils::vector_alloc<float, FloatVectorAlignment, FloatVectorWidth>(
+                    settings.max_samples * Configuration::num_scalars_rounded),
+                &free);
+
+            const auto buffer_index = [&buffer](std::size_t index) -> float *
+            { return buffer.get() + index * Configuration::num_scalars_rounded; };
+
+            std::vector<std::size_t> parent(settings.max_samples);
+            std::vector<std::vector<std::size_t>> children(settings.max_samples, std::vector<std::size_t>());
+            std::vector<float> cost(settings.max_samples, -1.0);
+
+            const auto add_edge = [&parent, &children](std::size_t parent_node, std::size_t child_node)
+            {
+                parent[child_node] = parent_node;
+                children[parent_node].emplace_back(child_node);
+            };
+            const auto remove_edge = [&parent, &children](std::size_t parent_node, std::size_t child_node)
+            {
+                parent[child_node] = -1;
+                auto it = std::find(children[parent_node].begin(), children[parent_node].end(), child_node);
+                if (it != children[parent_node].end())
+                {
+                    children[parent_node].erase(it);
+                }
```

Comment:
No need to change right now, but worried about performance with `erase` - that might cause a lot of memory copies if children are early in the list.